formatter/src/main/java/org/togetherjava/formatter/Formatter.java

```diff
@@ -0,0 +1,319 @@
+                        || (lastToken == TokenType.BIGGER && type != TokenType.OPEN_PARENTHESIS)
+                        || lastToken == TokenType.RETURN || lastToken == TokenType.COMMA
+                        || lastToken == TokenType.CATCH
+                        || (lastToken == TokenType.CLOSE_PARENTHESIS
+                                && type != TokenType.CLOSE_PARENTHESIS && type != TokenType.DOT
+                                && type != TokenType.COMMA)
+                        || lastToken == TokenType.PACKAGE) {
+                    result.append(" ");
+                }
+
+                if (!forClosed) {
+                    if (type == TokenType.OPEN_PARENTHESIS) {
+                        forClosingLevel++;
+                    } else if (type == TokenType.CLOSE_PARENTHESIS) {
+                        forClosingLevel--;
+                    }
+
+                    if (forClosingLevel == 0) {
+                        forClosed = true;
+                    }
+                }
+
+                if (type == TokenType.FOR) {
+                    forClosed = false;
+                }
+
+                if (!(NON_SPACE_TOKENS.contains(type) || NON_SPACE_TOKENS.contains(lastToken))) {
+                    result.append(" ");
+                }
+
+                append(result, token);
+            }
+
+            lastToken = type;
+        }
+
+        return new StringBuilder(result.toString().replaceAll("([;}])\n\n", "$1\n"));
+    }
+
+    /**
+     * Puts the needed indentation into a StringBuilder
+     *
+     * @param indentation indentation level
+     * @param sb string builder
+     * @author illuminator3
+     */
+    private void putIndentation(int indentation, StringBuilder sb) {
+        IntStream.range(0, indentation).forEach(n -> sb.append("    "));
+    }
+
+    /**
+     * Appends a token to a StringBuilder
+     *
+     * @param sb string builder
+     * @param token token to append
+     * @author illuminator3
+     */
+    private void append(StringBuilder sb, Token token) {
+        sb.append(token.content());
+    }
+
+    /**
+     * Makes a list mutable
+     *
+     * @param in (im)mutable list
+     * @return mutable list
+     * @author illuminator3
+     */
+    private <T> List<T> makeMutable(List<T> in) {
+        return new ArrayList<>(in);
+    }
+
+    /**
+     * Removes every whitespace from a given list of tokens
+     *
+     * @param tokens tokens to remove whitesapces from
+     * @author illuminator3
+     */
+    private void purgeWhitespaces(List<Token> tokens) {
+        tokens.removeIf(t -> t.type() == TokenType.WHITSPACE);
+    }
+
+    /**
+     * Indexes tokens to contain information about whether they are code tokens or not
+     *
+     * @param tokens not-indexed tokens
+     * @return indexed tokens
+     * @author illuminator3
+     */
+    private List<IndexedToken> indexTokens(List<Token> tokens) {
```

Comment:
This might be easier to read as
```java
return tokens.stream()
             .map(token -> new IndexedToken(token, isCodeToken(token)))
             .toList();
```